### PR TITLE
Added template function to generate AMI id

### PR DESCRIPTION
**Pull request by arjunrn**
Added a new `amiID` function to the template context which can get the AMI ID when given the `name` and `owner-id`.

### CODE REVIEW

File: provisioner/template.go [hunk 1]
```diff
@@ -270,6 +275,11 @@ func renderTemplate(context *templateContext, filePath string) (string, error) {
 		"publicKey":                 publicKey,
 		"stupsNATSubnets":           stupsNATSubnets,
 	}
+	if len(extraTemplateFuncs) > 0 {
```
Comment (linki) [inline provisioner/template.go]: I think we can drop the if here.

File: provisioner/node_pools.go [hunk 2]
```diff
@@ -64,6 +69,38 @@ type AWSNodePoolProvisioner struct {
...
+		ownerAccount := params[1]
+		input := ec2.DescribeImagesInput{Filters: []*ec2.Filter{
+			{Name: &filterNameName, Values: []*string{&name}},
+			{Name: &filterOwnerName, Values: []*string{&ownerAccount}},
+		}}
+		output, err := a.ec2Client.DescribeImages(&input)
+		if err != nil {
+			return "", fmt.Errorf("failed to describe image with name %s and owner %s: %v", name, ownerAccount, err)
+		}
+		if len(output.Images) != 1 {
```
Comment (linki) [inline provisioner/node_pools.go]: When search for AMIs by name and owner, isn't it expected that we get multiple back from AWS (and then select the most recent one?)

Comment (arjunrn) [reply provisioner/node_pools.go]: Yes in the normal case but I am assuming here that the image names are immutable because of the pipelines.

Comment (linki) [reply provisioner/node_pools.go]: So it's something like `kubernetes-image-ami-master-42`?

**Comment by aermakov-zalando** [comment]
Thanks! Is there any reason for doing it only in one case, or was it just simpler to write? I'd prefer all templates to work the same, i.e. add `amiID` function to the default list and just pass the AWS client to `templateContext`. How hard would it be?

File: provisioner/node_pools.go (hunk 2 above)
Comment (arjunrn) [reply provisioner/node_pools.go]: Yes

File: provisioner/template.go [hunk 3]
```diff
@@ -552,3 +558,24 @@ func stupsNATSubnets(vpcCidr string) ([]string, error) {
...
+func amiID(adapter *awsAdapter, imageName, imageOwner string) (string, error) {
+	if adapter == nil || adapter.ec2Client == nil {
+		return "", fmt.Errorf("the ec2 client is not available")
+	}
+
+	filterNameName := describeImageFilterNameName
+	filterOwnerName := describeImageFilterNameOwner
+	input := ec2.DescribeImagesInput{Filters: []*ec2.Filter{
+		{Name: &filterNameName, Values: []*string{&imageName}},
+		{Name: &filterOwnerName, Values: []*string{&imageOwner}},
```
Comment (linki) [inline provisioner/template.go]: You could also use `aws.String()` for that. You might be able to drop the temporary variables as well. There's also [StringSlice](https://godoc.org/github.com/aws/aws-sdk-go/aws#StringSlice) for the values part.

File: provisioner/template_test.go [hunk 4]
```diff
@@ -507,3 +514,72 @@ func TestStupsNATSubnetsErrors(t *testing.T) {
...
+		imageID   string
+		output    []*ec2.Image
+		expectErr bool
+	}{
+		{
+			name:      "basic",
+			imageName: "kubernetes-image-ami",
+			ownerID:   "8085",
+			imageID:   "ami-0001dsf",
+			output:    []*ec2.Image{{ImageId: stringPtr("ami-0001dsf")}},
```
Comment (linki) [inline provisioner/template_test.go]: Let's use `aws.String()` like [here](https://github.com/kubernetes-incubator/external-dns/blob/master/provider/aws.go#L485).

**Comment by aermakov-zalando** [comment]
Do you think it makes sense to cache the resolved AMIs (that would most likely be the same in different node pool templates) somewhere, e.g. in the template context? This might help with the rate limits in clusters with a large number of node pools.

**Comment by aermakov-zalando** [comment]
LGTM otherwise, but please use `aws.String/aws.StringValue` wrappers as @linki suggested.

File: provisioner/template.go [hunk 5]
```diff
@@ -84,14 +89,15 @@ type podResources struct {
...
-func newTemplateContext(baseDir string, cluster *api.Cluster, nodePool *api.NodePool, values map[string]interface{}, userData string) *templateContext {
+func newTemplateContext(baseDir string, cluster *api.Cluster, nodePool *api.NodePool, values map[string]interface{}, userData string, adapter *awsAdapter) *templateContext {
 	return &templateContext{
 		baseDir:      baseDir,
 		manifestData: make(map[string]string),
 		cluster:      cluster,
 		nodePool:     nodePool,
 		values:       values,
 		userData:     userData,
+		awsAdpater:   adapter,
```
Comment (aermakov-zalando) [inline provisioner/template.go]: Typo: _awsAdapter_.

**Comment by aermakov-zalando** [comment]
👍

**Comment by arjunrn** [comment]
:+1:

**Comment by aermakov-zalando** [comment]
👍

**Pull request merged**